Tiny (internal) Wallet

KtorZ opened the pull request:
Still WIP, but the essence is there. Only remains to implement the `applyBlock` function which maintains the UTXO state. This one is a good candidate for some property-based testing :blush:

KtorZ committed:
Factor out some common code in a Hydra.Chain.Direct.Util module.

KtorZ committed:
Implement a tiny wallet which follows the chain and keep track of a UTXO at a single address.

  Still WIP, but it's a good head start. It relies on the mini-protocols as well to watch the chain and maintain its state.

abailly commented:
Why the name `Wallet` instead of simply `Client`? Perhaps this is something you already discussed.

KtorZ committed:
Write property-tests for 'applyBlock' of the tiny wallet.

KtorZ committed:
Implement 'applyBlock'.

KtorZ committed:
Implement mock state-query client.

KtorZ committed:
Move the Ouroboros mini-protocols mock server in its own (test) module

KtorZ committed:
Remove redundant temporary mock server in 'DirectSpec'.

KtorZ committed:
Write basic tests for 'withTinyWallet'

  - I've changed a bit the 'TinyWallet' API to work with STM actions instead. It's much nicer than to work in the base monad, as we can retry STM actions.
  - I also slightly tweaked the mock server setup so it'll create the socket in a temp directory and use an arbitrary magic, so it's less boilerplate for those consumming it downstream.

KtorZ commented:
@abailly because we have two clients already. One client is the `Direct` chain component, and the other is this `TinyWallet`. That the wallet is an Ouroboros client here is however _an implementation detail_. This could be implemented by making API calls to the cardano-wallet in the end and it wouldn't matter for the `Direct` chain component. Hence, I prefer to call that by its function (i.e. Wallet) than by how its realized.

abailly commented:
Makes sense. From the looks of it I thought this was actually the chain sync client for the `Direct` chain component.
